### PR TITLE
Align the scheduling name with the instruction name after HLO rematerialization

### PR DESCRIPTION
**Pull request by zhenying-liu**
This CL is to fix the assertion failures in HLO verifier when checking scheduling names after HLO rematerialization.

### CODE REVIEW

**Review by cheshire — CHANGES_REQUESTED** [review]
Thanks ! I don't fully follow: which test was previously triggering the assertion failure?

**Comment by zhenying-liu** [comment]
> Thanks ! I don't fully follow: which test was previously triggering the assertion failure?

My coworker ran PAXML and found the assertion failure that worked before. This was due to a recent change in HLO verifier that disallowing the mismatch between the scheduling name and the instruction name.

**Comment by zhenying-liu** [comment]
I found some unit test failures like xla/service/hlo_module_test.cc. I'm working on that.

**Review by nouiz — COMMENTED** [review]
Does the HloVerifier runs in all tests inside hlo_remat..._test.cc?
If not, can you add that? To be sure that it always pass.

Can you confirm the updated tests would fail without the fix?

**Comment by zhenying-liu** [comment]
> Does the HloVerifier runs in all tests inside hlo_remat..._test.cc? If not, can you add that? To be sure that it always pass.
> 
> Can you confirm the updated tests would fail without the fix?

Yes, I confirm the updated tests would fail without the fix. For example, the add instruction names in below computations are

- computation: %add_mul_comp, add: %add
- computation: %add_mul_comp_clone, add: _%add.3_ 
- computation: %entry,  add: %add.1 and _%add.2_

There are two changes in hlo_rematerialization.cc. 

- The first change is to update scheduling names for instructions in the **cloned** node: %add_mul_comp_clone. 
- And the second change update scheduling names mostly for instructions in the %entry.

File: xla/service/hlo_rematerialization_test.cc [hunk 1]
```diff
@@ -982,7 +995,7 @@ ENTRY %mycomp (param: f32[1]) -> f32[1] {
   // Original all-gather.
   const HloInstruction* original_ag = FindInstruction(module.get(), "ag");
   // The all-gather should have been rematerialized
-  const HloInstruction* remat_ag = FindInstruction(module.get(), "ag.remat");
+  const HloInstruction* remat_ag = FindInstruction(module.get(), "ag.remat2");
```
[the remark this answers is not in the document]
Comment (zhenying-liu) [reply xla/service/hlo_rematerialization_test.cc]: I removed this change as I took back the changes in hlo_instruction.cc regarding cloning: the cloned instruction will add the context suffix to the end of the instruction name and cause duplicated names to be renamed as ag.remat2.
The Clone() never takes the context suffix name but I won't fix it in this CL.

File: xla/hlo/ir/hlo_instruction.cc [hunk 2]
```diff
@@ -2654,7 +2654,16 @@ std::unique_ptr<HloInstruction> HloInstruction::CloneWithNewOperands(
   // The new instruction's name will be uniquified when it's added to a
   // computation.
   clone->SetAndSanitizeName(name());
+  // Add the context's suffix name to the cloned instruction
+  if (context != nullptr && !context->suffix().empty()) {
```
[the remark this answers is not in the document]
Comment (zhenying-liu) [reply xla/hlo/ir/hlo_instruction.cc]: I won't change hlo_instruction.cc for now to simplify this CL.

**Review by nouiz — APPROVED** [review]

**Comment by shawnwang18** [comment]
Hello, the issue solved by this PR is breaking many runs, could we prioritize it and make it fixed ASAP?

**Review by akuegel — APPROVED** [review]

**Comment by nouiz** [comment]
A Google internal crash. Anything we can do from outside Google?

**Comment by zhenying-liu** [comment]
I'm not sure about the current **Bazel Tags/bazel-tags** failure. I don't think it's caused by my change.
```
Traceback (most recent call last):
  File "/home/runner/work/xla/xla/build_tools/lint/tags.py", line 128, in <module>
    main()
  File "/home/runner/work/xla/xla/build_tools/lint/tags.py", line 114, in main
    raise ValueError(
ValueError: Tag(s) {'config-cuda-only'} are undocumented! Please document them in `build_tools/lint/tags.py`.
```

**Comment by zhenying-liu** [comment]
> A Google internal crash. Anything we can do from outside Google?

I made the check stricter: only update scheduling_name for those mismatched instructions; don't change constants' instruction names as they won't be checked.

**Comment by zhenying-liu** [comment]
> Hello, the issue solved by this PR is breaking many runs, could we prioritize it and make it fixed ASAP?

@shawnwang18 You mentioned the failure in SanitizeConstantNames. I tried to address it in [PR 16241](https://github.com/openxla/xla/pull/16241), but I was pointed out that https://github.com/openxla/xla/commit/ca5545b2585e151b243e1b94b39f0ff126536e35 already ignored the constants in HLO verifier. If you included this commit, it shouldn't be a problem any more.

**Comment by zhenying-liu** [comment]
@sgerrard @cheshire @akuegel Please help with another code review. Thanks.

**Comment by akuegel** [comment]
@golechwierowicz this PR shows what kind of name changes are happening in HloRematerialization pass. Since we have already disabled the verifier, this PR is possibly obsolete, but I also am not sure whether just "hiding" the rename from the verifier would have been the right fix?

**Comment by golechwierowicz** [comment]
Hiding the rename is not the right solution.

Ultimately remat introduces new instructions via cloning. We should make sure they're created with a .remat or .clone suffix. See this piece of code: https://github.com/openxla/xla/blob/91ecf95a08bcc5d5b0b63c5db15934f1d1c51a2c/xla/service/hlo_verifier.cc#L2139-L2148.

In this case `convert.2367.2` should be something like `convert.2367.1.remat`.

IMO the remat pass should still be fixed.

**Comment by cheshire** [comment]
Alternatively, maybe we could consider disabling XLA remat entirely, at least from GPUs? Consistent feedback from @hawkinsp is that it generally makes things worse, and there are better mechanisms to do it explicitly using JAX APIs.

**Comment by akuegel** [comment]
As far as I understood, this PR as is became obsolete. @golechwierowicz confirmed it is the wrong fix, and we have disabled the verifier to unblock users.

**Pull request closed without merging**